### PR TITLE
Prevent worker respawns from blocking container restarts

### LINKED ISSUE #1585: MainThread does not exit after "Exiting Harper process to trigger a container restart" log
### Bug Summary

After emitting the exit warning log, the Harper MainThread process fails to terminate, preventing the container from restarting. Domain socket failure results.

### Reproduction Repository (if applicable)

_No response_

### Steps to Reproduce

1. Install a component on a Fabric-managed node
2. Harper emits: [main/0] [warn]: Exiting Harper process to trigger a container restart
3. Observe that the MainThread process does not exit
4. Container remains up (uptime unaffected); domain socket enters failed state
5. Health and status checks begin failing
6. Node recovers only after a manual restart

### Expected Behavior

After emitting the exit warning, the Harper process should terminate cleanly, 
allowing the container runtime to perform an automatic restart.

### Actual Behavior

The MainThread process does not exit after the warning is logged. The container 
remains running with an 8-day uptime confirmed. The domain socket enters a failed 
state, causing health and status check failures across the affected node. Manual 
restart required to recover.

### Platform

harperfast/harper-pro:5.1.12

### Console Errors

[main/0] [warn]: Exiting Harper process to trigger a container restart

```
26-07-03T00:21:55.844110768Z"}
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
⨯ unhandledRejection:  Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: unhandledRejection in worker thread 33: Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: unhandledRejection Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
unhandledRejection Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: process.exit(1) called in worker thread 33 — ignored to keep Harper alive. Use process._realExit() for legitimate internal shutdown.
Error: process.exit() intercepted
    at process.interceptedExit [as exit] (/home/harperdb/.npm-global/lib/node_modules/@harperfast/harper-pro/dist/core/server/threads/workerProcessGuard.js:97:26)
    at /home/harperdb/harper/components/JJill-v5/node_modules/next/dist/lib/setup-exception-listeners.js:8:13
    at process.filteringUnhandledRejectionHandler (/home/harperdb/harper/components/JJill-v5/node_modules/next/dist/server/node-environment-extensions/unhandled-rejection.external.js:491:17)
    at process.emit (node:events:509:28)
    at emitUnhandledRejection (node:internal/process/promises:252:13)
    at throwUnhandledRejectionsMode (node:internal/process/promises:388:19)
    at processPromiseRejections (node:internal/process/promises:475:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:106:37)
Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
⨯ unhandledRejection:  Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: unhandledRejection in worker thread 33: Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: unhandledRejection Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
unhandledRejection Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: process.exit(1) called in worker thread 33 — ignored to keep Harper alive. Use process._realExit() for legitimate internal shutdown.
Error: process.exit() intercepted
    at process.interceptedExit [as exit] (/home/harperdb/.npm-global/lib/node_modules/@harperfast/harper-pro/dist/core/server/threads/workerProcessGuard.js:97:26)
    at /home/harperdb/harper/components/JJill-v5/node_modules/next/dist/lib/setup-exception-listeners.js:8:13
    at process.filteringUnhandledRejectionHandler (/home/harperdb/harper/components/JJill-v5/node_modules/next/dist/server/node-environment-extensions/unhandled-rejection.external.js:491:17)
    at process.emit (node:events:509:28)
    at emitUnhandledRejection (node:internal/process/promises:252:13)
    at throwUnhandledRejectionsMode (node:internal/process/promises:388:19)
    at processPromiseRejections (node:internal/process/promises:475:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:106:37)
Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
⨯ unhandledRejection:  Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[main/0] [notify]: Restarting Harper. This may take up to 60 seconds.
[job/33] [error]: unhandledRejection in worker thread 33: Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: unhandledRejection Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
unhandledRejection Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[job/33] [error]: process.exit(1) called in worker thread 33 — ignored to keep Harper alive. Use process._realExit() for legitimate internal shutdown.
Error: process.exit() intercepted
    at process.interceptedExit [as exit] (/home/harperdb/.npm-global/lib/node_modules/@harperfast/harper-pro/dist/core/server/threads/workerProcessGuard.js:97:26)
    at /home/harperdb/harper/components/JJill-v5/node_modules/next/dist/lib/setup-exception-listeners.js:8:13
    at process.filteringUnhandledRejectionHandler (/home/harperdb/harper/components/JJill-v5/node_modules/next/dist/server/node-environment-extensions/unhandled-rejection.external.js:491:17)
    at process.emit (node:events:509:28)
    at emitUnhandledRejection (node:internal/process/promises:252:13)
    at throwUnhandledRejectionsMode (node:internal/process/promises:388:19)
    at processPromiseRejections (node:internal/process/promises:475:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:106:37)
Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
⨯ unhandledRejection:  Error: Client network socket disconnected before secure TLS connection was established
    at ignore-listed frames {
  code: 'ECONNRESET',
  path: undefined,
  host: 'ns7-us-lax-2.hcl-qa2-auth.jjill-nonprod.harperfabric.com',
  port: '9933',
  localAddress: undefined
}
[main/0] [warn]: Exiting Harper process to trigger a container restart
```

### Screenshots or Videos

_No response_

### Additional Context

Log export confirms the exit warning fired across 6 hosts spanning 2 clusters 
and 3 regions between 00:10 and 00:27 UTC on 2026-07-03, totalling ~45 log entries.

On at least 2 hosts, the exit warning was preceded by repeated ECONNRESET errors 
on port 9933 and a worker thread calling process.exit(1), which was intercepted 
and suppressed by workerProcessGuard.js. The replication WebSocket also dropped 
(code 1006) immediately before the exit warning on one of the affected hosts.

The workerProcessGuard suppression of process.exit(1) in worker threads may be 
a contributing factor -- if the MainThread exit path relies on a similar mechanism 
or is blocked waiting on worker thread state, it may silently no-op under these 
conditions.

### Are you planning to fix this issue?

No, just reporting the issue

### First-time contributor support

- [ ] I'm new to contributing and would appreciate guidance on the process
- [ ] I'd like help understanding the project structure
- [ ] I need assistance with setting up the development environment
- [ ] I'm comfortable contributing, but new to this project specifically

### PR DESCRIPTION
**Pull request by kriszyp**
Terminal process shutdown now latches [worker creation](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-02c28577c2ca63129ed10809ce3db1469c70bef080e04fd273f8cbd510b6521bR330) and every replacement path, preventing [a late worker exit](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-02c28577c2ca63129ed10809ce3db1469c70bef080e04fd273f8cbd510b6521bR446) or an overlapping rolling restart from resurrecting the pool during teardown. The restart path [sets that latch as its first act](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-79b2936f95589c02e1c11589c45b5496220455fbcdd94fc29181cccdcb2f55a4R75) rather than waiting for `shutdownWorkersNow()`, so a debounced component reload cannot slip in and pre-start an HTTP replacement into a process that is already exiting. Container restarts also gain a bounded post-compaction exit watchdog, with an init process that can reap and restart a force-killed Harper child while preserving compatibility with volumes written by PID-1 images. Shutdown listeners now [preserve requested failure exit codes and still terminate when PID cleanup fails](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-01680253cab1229d8cdc4eff0a56337accf055e871c4ab4d87cb5b4879590445R66).

Every way the watchdog's shell can give up is a silent `exit 0` — an unreadable procfs, a base image without `sleep` — so a successful spawn was not evidence it would fire. It now [emits a readiness token](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-858756f15bb7ea5fb9b81127ae62a682bd55d55d4d25e5fffe87745e18d80c4fR9) once both facilities are proven, and [arming reports failure unless that token arrives](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-858756f15bb7ea5fb9b81127ae62a682bd55d55d4d25e5fffe87745e18d80c4fR69) — so the "restart teardown is unbounded" error reaches operators in exactly the environments that need it.

## For the human reviewer

1. The published image now [runs `tini -g` as PID 1](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R111), rather than Harper itself. This makes a reliable SIGKILL watchdog possible and gives the container a reaper, but group-forwarding also changes what signals component subprocesses receive on `docker stop`. [`DESIGN.md` now records the user-visible consequences](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-3dc5dd454e080eb849ee5efaf79df2585fbe0a06804d69249b4c81da05a63875R827), including that `-g` does *not* reach Harper-managed subprocesses (`processManagement` forks them `detached: true`, which `setsid()`s them out of the group). Reverting is localized to the Docker entrypoint; rejecting it removes the hard-exit guarantee selected for this issue. This process-group boundary is where to look hardest.
2. The watchdog is an external procfs-aware shell child rather than a Node timer or supervisor-owned timeout. It fails closed on the facilities it can check and fails open when parentage cannot be verified; arming failure is reported as an unbounded teardown. The child deliberately inherits stderr so its force-kill diagnostic survives a wedged Harper event loop. That can delay EOF by up to 60 seconds for an unusual bare-metal supervisor using a piped stderr; the shipped `tini` container is unaffected. Replacing the helper later is localized.
3. The watchdog starts [after configured storage compaction](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-79b2936f95589c02e1c11589c45b5496220455fbcdd94fc29181cccdcb2f55a4R66), rather than bounding the entire restart. This avoids SIGKILL during live database-directory rename operations; changing the boundary is mechanically small, but a different decision needs an explicit recovery contract for interrupted compaction. Deliberately *not* changed in this round: the latch stays below `compactOnStart()`, because latching across a multi-minute compaction would also disable the unexpected-exit auto-restart while the node is still taking traffic.
4. The terminal latch is one-way and process-global, while scoped worker shutdowns do not engage it. This favors simple teardown correctness over lifecycle generations or cancellation and is the broadest coupling in the change; changing it later would touch every worker-start path. One consequence a reviewer should weigh: `components/componentLoader.ts`'s `runRestartCycle()` has no way to read the latch, so under `harper dev` (`autoReload`, the only caller that subscribes to `requestRestart`) a debounced cycle can still run `loadComponentDirectories()` during teardown. The `restartWorkers()` half is latched, so nothing respawns; a reader on `manageThreads` would close the reload half.
5. Docker verification composes clean exit, stale-PID reboot, self-exit, and force-kill fault injection as separate paths rather than one request-to-orchestrator-resurrection scenario. A single end-to-end resurrection leg would strengthen coverage but add workflow complexity.
6. Two review suggestions were declined. Adding a `loadRootComponents` spy to the late-restart harness would pin statement ordering rather than a failure mode. The `uncaught exception → exit 1` half of the exit-contract change is not testable as stated, and does not hold: `server/threads/socketRouter.ts` installs a main-thread `uncaughtException` handler that logs and continues, so an uncaught throw never reached the `'exit'` listener's code either way. The exit-code change is real for *requested* codes, which is what the new tests cover.
7. Cross-model reviewers repeatedly flagged the added comments as narrating control flow. They have been compressed twice; the surviving one-liners exist because a PR reviewer asked for them (notably the `bin/run.ts` forward-defence note). Cutting them further is a style call worth making explicitly rather than silently.

## Verification

- `npm run build` and `npm run lint:required` — passed on the final head.
- `npm run test:unit:bin` — 248 passing, 7 pending on the final head (241 before this branch).
- [`unitTests/server/threads/terminalShutdown.test.js`](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-73d8cf4523ec3c1c94169ea7215f3d5d0dac53b7d1c27317c11e755100836bf9R76) — 5 passing: concurrent rolling restart, unexpected exit, non-overlapping replacement, a restart requested *after* the latch, and scoped shutdown. The new late-restart case is mutation-verified: deleting the early return in `restartWorkers()` turns it red on `restartNumber`.
- [`unitTests/bin/exitListeners.test.js`](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-06bf3e2599e2e799c8321091e567af43e1a028157efde4e5f69a6376670bc3ceR17) — 3 passing: a requested non-zero code survives the `'exit'` listener, zero survives, and SIGTERM still exits 0.
- [`unitTests/bin/restartExitWatchdog.test.js`](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-f39aafa5803a99c592f429801fbfd46ec34bc5a11248be77a8610411bf34f76aR49) — 6 passing, covering the script withholding its token without `sleep`, and the caller's disarm wiring for token / early exit / readiness timeout.
- `npm run test:integration -- integrationTests/server/rolling-restart.test.ts integrationTests/components/shutdown-drain-e2e.test.ts` — 3 passing on the final head.
- Reproduced the Docker fault injection locally against the built `dist/`: arming completes its handshake, the event loop is then wedged with `Atomics.wait`, and the process dies with exit 137 and the watchdog diagnostic after 1 second. The [smoke workflow's fault-injection leg](https://github.com/HarperFast/harper/pull/2316/changes?w=1#diff-302c8b03f1c280489dd05fe5bd939545771c0057f1b1e367474b415b0c253167R225) was updated for the now-async arming; the container legs themselves run in CI on this PR.

Closes HarperFast/harper#1585

Complexity: complicated

— Codex (GPT-5.6), with review-feedback rounds by Claude Opus 5

_Generated by Claude Opus 5._

<sub>Review-Coverage: authored=claude; ran=codex,gemini; declined=cursor-grok,cursor-composer,domain; rounds=5 @ f0dc7191177b</sub>

<sub>Human-Review-Need: 3 @ f0dc7191177b</sub>

### CODE REVIEW

File: bin/restartExitWatchdog.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+const WATCHDOG_SCRIPT = `
+parent_pid=$1
+delay_seconds=$2
+[ "$parent_pid" -gt 1 ] || exit 0
+# Shell PPID variables are snapshots; procfs reflects reparenting after Harper exits.
+read -r _ _ _ current_parent _ < "/proc/$$/stat" || exit 0
+[ "$current_parent" -eq "$parent_pid" ] || exit 0
+sleep "$delay_seconds" || exit 0
+read -r _ _ _ current_parent _ < "/proc/$$/stat" || exit 0
+[ "$current_parent" -eq "$parent_pid" ] || exit 0
```
[the remark this answers is not in the document]
Comment (kriszyp) [reply bin/restartExitWatchdog.ts]: Fixed in 1a62ecd31: both procfs parent checks now use POSIX string equality, so an empty or malformed value fails safely without integer-expression diagnostics.

File: bin/restartExitWatchdog.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		return false;
+	}
+
+	try {
+		const timeoutSeconds = Math.max(1, Math.ceil(timeoutMs / 1000));
+		const watchdog = spawn(
+			'/bin/sh',
+			['-c', WATCHDOG_SCRIPT, 'harper-restart-exit-watchdog', String(process.pid), String(timeoutSeconds)],
+			{
+				env: { PATH: '/usr/bin:/bin' },
```
[the remark this answers is not in the document]
Comment (kriszyp) [reply bin/restartExitWatchdog.ts]: Addressed in 4dad1d337 with a narrower environment boundary: /usr/bin and /bin remain first, the parent PATH is appended for non-FHS utility lookup, and every other parent environment variable remains scrubbed.

File: server/threads/manageThreads.js [hunk 3]
```diff
@@ -470,6 +477,7 @@ async function restartWorkers(
 	startReplacementThreads = true
 ) {
 	if (isMainThread) {
+		if (processShuttingDown && startReplacementThreads) return;
```
Comment (cb1kenobi) [inline server/threads/manageThreads.js:480]: **Medium: nothing tests a restart requested *after* the latch is set**

Mutation check: delete this line, rebuild (`tsc -p tsconfig.build.json`), and all four `terminalShutdown.test.js` cases still pass. Every harness mode sets the latch *during* an in-flight `restartWorkers()`; none calls `restartWorkers()` after `beginProcessShutdown()`.

That ordering is the more likely production one. `bin/restart.ts` latches first (via `shutdownWorkersNow()`), and the teardown that follows spans `closeServers()` + `cleanupChildrenProcesses()` — a component-watcher reload or a queued `requestRestart` landing in that window is exactly what resurrects the pool. This line is the only thing stopping it, and it is unguarded.

Suggested fix: add a harness mode that calls `beginProcessShutdown()` and *then* `restartWorkers('http')`, asserting no `onStarted` fires and `workers.length === 0`.

—
_Generated by Barber AI_

File: bin/run.ts [hunk 4]
```diff
@@ -51,21 +52,24 @@ function addUnhandleRejectionListener() {
...
-			fs.removeSync(path.join(env.get(terms.CONFIG_PARAMS.ROOTPATH), terms.HDB_PID_FILE));
-			process.exit(0);
+			try {
+				fs.removeSync(path.join(env.get(terms.CONFIG_PARAMS.ROOTPATH), terms.HDB_PID_FILE));
+			} catch (error) {
+				hdbLogger.error('Unable to remove the Harper pid file during shutdown', error);
+			}
 		};
 		process.on('exit', () => {
+			beginProcessShutdown();
```
Comment (cb1kenobi) [inline bin/run.ts:63]: **Medium: dropping `process.exit(0)` from the `'exit'` listener changes the container-visible exit contract, with no test and no release note**

The old listener body was `fs.removeSync(...); process.exit(0)`. Calling `process.exit(0)` from inside an `'exit'` listener rewrites the code for *every* exit, so this is a real behaviour change, not a refactor. Measured on Node 24:

| scenario | before | after |
|---|---|---|
| `process.exit(1)` | exit **0** | exit **1** |
| uncaught fatal exception | exit **0**, stack **suppressed** | exit **1**, stack printed |

The new behaviour is the correct one — a crashed Harper previously reported success and swallowed its own stack trace. But it is orchestrator-visible: `docker run --restart=on-failure` now restarts where it used to stay down, and a Kubernetes pod terminates as `Error` rather than `Completed`, which is what arms `CrashLoopBackOff`. Reverting `bin/run.ts` wholesale to the merge base leaves `unitTests/bin` at 240 passing / 0 failing, so nothing guards it.

Suggested fix: add a unit test asserting a non-zero code survives the listener (spawn a fixture that installs the listeners and calls `process.exit(1)`), and call the change out in the release notes.

Minor, same handler: the `beginProcessShutdown()` on this line has no observable effect. Both call sites are followed synchronously by `process.exit()` (or run inside `'exit'` itself, past the point of no return), so nothing can reach `startWorker()`/`restartWorkers()` in between. Worth a one-line comment saying it is forward-defence for a future async shutdown, otherwise it reads as load-bearing.

—
_Generated by Barber AI_

File: server/threads/manageThreads.js [hunk 5]
```diff
@@ -507,9 +515,10 @@ async function restartWorkers(
 		// listenOnPorts() treat a dedicated listener's EADDRINUSE as an external conflict.
 		const canPreStartReplacement = process.platform !== 'win32' && process.platform !== 'darwin' && !isBun;
 		for (let worker of workers.slice(0)) {
+			if (processShuttingDown && startReplacementThreads) break;
 			if ((name && worker.name !== name) || worker.wasShutdown) continue; // filter by type, if specified
 			const overlapping = OVERLAPPING_RESTART_TYPES.indexOf(worker.name) > -1;
-			if (overlapping && startReplacementThreads && canPreStartReplacement) {
+			if (overlapping && startReplacementThreads && canPreStartReplacement && !processShuttingDown) {
```
Comment (cb1kenobi) [inline server/threads/manageThreads.js]: **Low: `!processShuttingDown` here is unreachable, and it hides the fact that the Linux pre-start path is untested**

Line 518 already does `if (processShuttingDown && startReplacementThreads) break;`. Reaching this line means that was false, and this line requires `startReplacementThreads` — so `processShuttingDown` was false at 518. The only code in between (the `name`/`wasShutdown` filter and the `overlapping` lookup) is synchronous and side-effect-free, so the latch cannot flip. The condition is always true.

That matters beyond tidiness: this is the overlapping HTTP pre-start branch, i.e. the production path on Linux and the configuration in #1585. I forced `canPreStartReplacement = true` to simulate Linux and then deleted `&& !processShuttingDown` — all four tests still passed, because the branch has no coverage of its own either way.

Line 603 has the same shape; there the latch *can* in principle flip via a synchronous `worker.emit('shutdown')` listener, so it is defensible, but it is equally uncovered. Line 640 is the one that genuinely needs its guard (it runs inside an async `'exit'` handler) and it is the one the tests do kill.

Suggested fix: drop the redundant conjunct here and note in a comment that line 518 is the guard for this branch.

—
_Generated by Barber AI_

File: server/threads/manageThreads.js [hunk 6]
```diff
@@ -436,7 +443,7 @@ function startWorker(path, options = {}) {
 	});
 	worker.on('exit', (_code) => {
 		workers.splice(workers.indexOf(worker), 1);
-		if (!worker.wasShutdown && options.autoRestart !== false) {
+		if (!processShuttingDown && !worker.wasShutdown && options.autoRestart !== false) {
```
Comment (cb1kenobi) [inline server/threads/manageThreads.js:446]: **Low: merge-order coordination for this hunk, plus a note on the #2314 interaction**

Two other open PRs touch this file, and one touches these exact lines:

- [#2129 "Fix HTTP startup after pre-ready worker restart"](https://github.com/HarperFast/harper/pull/2129) rewrites the `else` branch two lines below this one (adds `options.onRestartExhausted?.(worker)` and fixes the `worker.restarts` → `worker.unexpectedRestarts` message). Git will probably auto-merge, but whoever lands second should confirm both survive: the `!processShuttingDown &&` conjunct on this `if`, and the new `else` body.
- [#2085 "avoid startup hangs on zombie process groups"](https://github.com/HarperFast/harper/pull/2085) adds an entry to the same `module.exports` block where this PR adds `beginProcessShutdown,` (line 155) — a near-certain textual conflict, trivial to resolve.

Worth stating explicitly for the reviewer: [#2314 "Keep a pre-ready worker's event loop alive through startup"](https://github.com/HarperFast/harper/pull/2314) adds a 600s `startupDeadline` that calls `realExit(1)`, and the auto-restart on this line is what would turn a lost `clearTimeout` into a kill loop. This PR does **not** change that — `processShuttingDown` is false during normal operation, so the new conjunct short-circuits to the old behaviour. Neither mitigated nor worsened; just don't read the new latch as covering it.

—
_Generated by Barber AI_

File: bin/restart.ts [hunk 7]
```diff
@@ -56,30 +57,40 @@ async function restart(req: any) {
...
-			unlinkSync(path.join(envMgr.get(hdbTerms.CONFIG_PARAMS.ROOTPATH), hdbTerms.HDB_PID_FILE));
-			hdbLogger.debug('Starting new process...');
-			if (process.env.HARPER_EXIT_ON_RESTART) {
-				// use this to exit the process so that it will be restarted by the
-				// PM/container/orchestrator.
-				hdbLogger.warn('Exiting Harper process to trigger a container restart');
-				process.exit(0);
+			try {
+				if (process.env.HARPER_EXIT_ON_RESTART && !armRestartExitWatchdog(hdbTerms.RESTART_TIMEOUT_MS))
+					hdbLogger.error('Restart exit watchdog is unavailable; restart teardown is unbounded');
```
Comment (cb1kenobi) [inline bin/restart.ts:82]: **Nit: error-level log for a condition that can never be satisfied off Linux**

`armRestartExitWatchdog()` returns `false` immediately on non-Linux, having already logged `warn('Restart exit watchdog is only available on Linux')`. This line then logs the same fact again at `error`. On a macOS or Windows host with `HARPER_EXIT_ON_RESTART` set, every restart emits an error-severity line for a permanent, unactionable platform condition — and error-level logs feed alerting on Fabric.

Suggested fix: only escalate when the platform could have supported it.

```suggestion
				if (process.env.HARPER_EXIT_ON_RESTART && !armRestartExitWatchdog(hdbTerms.RESTART_TIMEOUT_MS) && process.platform === 'linux')
```

—
_Generated by Barber AI_

File: Dockerfile [hunk 8]
```diff
@@ -105,6 +107,7 @@ EXPOSE 9926
 EXPOSE 9932
 EXPOSE 9933
 
-ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh"]
+# Harper must not be PID 1 so its restart watchdog can force a wedged teardown to exit.
+ENTRYPOINT ["/usr/bin/tini", "-g", "--", "/usr/local/bin/docker-entrypoint.sh"]
```
Comment (cb1kenobi) [inline Dockerfile:111]: **Nit: PID 1 changing from Harper to `tini` is user-visible and undocumented**

`server/DESIGN.md` gained a paragraph on `beginProcessShutdown()`, but nothing records that the published image no longer runs Harper as PID 1. That is observable to users — `docker exec … ps`, anything keying off PID 1, `docker run --init` now nesting two inits, and `-g` group-forwarding changing which processes receive `docker stop`'s SIGTERM. The PR body itself calls the process-group boundary the thing to look hardest at, which is a good argument for it being written down somewhere durable.

Suggested fix: a release-note entry, and a line in the Docker section of the docs.

Also unexercised: the new smoke leg asserts `tini` is PID 1 and that `docker stop` exits 0, but against a container with no components installed and therefore no child processes — so the `-g` forwarding behaviour has no coverage at all. Adding a component that forks a child before the `docker stop` would close that.

—
_Generated by Barber AI_
